### ISSUE TITLE
Error parsing telegram

### ISSUE THREAD

**Issue by elucidator** [nielsbasjes/dsmr-tools#54]
When i try to parse the input delivered from from my P1 endpoint i get the error:

```
java.lang.NullPointerException: Cannot invoke "org.antlr.v4.runtime.Token.getText()" because "ctx.timestamp" is null

	at nl.basjes.dsmr.ParseDsmrTelegram.visitMBus2Usage(ParseDsmrTelegram.java:362)
	at nl.basjes.dsmr.ParseDsmrTelegram.visitMBus2Usage(ParseDsmrTelegram.java:92)
	at nl.basjes.dsmr.parse.DsmrParser$MBus2UsageContext.accept(DsmrParser.java:639)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at nl.basjes.dsmr.ParseDsmrTelegram.visitTelegram(ParseDsmrTelegram.java:239)
	at nl.basjes.dsmr.ParseDsmrTelegram.parse(ParseDsmrTelegram.java:156)
	at nl.basjes.dsmr.ParseDsmrTelegram.parse(ParseDsmrTelegram.java:118)
	at nl.basjes.dsmr.parse.TestDsmrParser.myTest(TestDsmrParser.java:298)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at <SNIP>
```
it seems that is crashing at the processing of the gasData, i am unfamiliar with the antlr4 parsing method, but i can t  seem to find the error. Happy to help if you could provide some pointers.

```
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16719.940*kWh)
1-0:1.8.2(19403.220*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.057*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.89*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211122210000)(00)(60)(1)(0-2:24.2.1)(m3)
(13368.864)
0-2:24.4.0(1)
!
```

from what I understand is that the strings:
```
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211122220000)(00)(60)(1)(0-2:24.2.1)(m3)
(13370.389)
```

Are formatted differently that the your g4 spec defines, the key 0-2:24.2.1 is not on a seperate line, but instead follows directly after 0-2:24.3.0

Could this be due to an older meter?

**Comment by nielsbasjes** [comment]
What I found is that this seems to be something that is not part of the official specifications but does occur in several places.

I did find more examples on the internet that essentially contain fragments similar to what you found. For example this:
```
0-2:24.1.0(3)
0-2:96.1.0(3238313031353431303034333539373131)
0-2:24.3.0(160129230000)(00)(60)(1)(0-2:24.2.1)(m3)
(09871.759)
```

Now some of this I can figure out myself (like the timestamp) but several parts I really need some kind of specification what it all means (mainly the `(00)(60)(1)` part).

So if you have any kind of specification for me that would be really useful.

To check: In your case 
```
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211122210000)(00)(60)(1)(0-2:24.2.1)(m3)
(13368.864)
```

does it mean that on 2021-11-22 at 21:00:00 (`(211122210000)`) your gas meter (`0-2:24.1.0(3)` --> type 3 = gas) reported 13368.864 M^3 (`(13368.864)` `(m3)`) ?

**Comment by nielsbasjes** [comment]
Request: Please provide me a full message (or even better; 3 or 4 messages). 
I expect that after the '!' at the end there is a CRC code.
If there is not then that is also something my parse must be able to deal with.

**Comment by nielsbasjes** [comment]
I think I fixed it.
I would really appreciate it if you could do some tests before I release it.

You can get version 0.6-SNAPSHOT from https://oss.sonatype.org/content/repositories/snapshots/

**Comment by elucidator** [comment]
> So if you have any kind of specification for me that would be really useful.
Have been looking around as well but can not find anything on that part. 
Funny thing thou when you have a look at the 3.0 spec and haven a look at the example, it shows the data as i recieve it.
(http://domoticx.com/p1-poort-slimme-meter-hardware/)


> 
> To check: In your case
> 
> ```
> 0-2:24.1.0(3)
> 0-2:96.1.0(3238303131303038333036343239303133)
> 0-2:24.3.0(211122210000)(00)(60)(1)(0-2:24.2.1)(m3)
> (13368.864)
> ```
> 
> does it mean that on 2021-11-22 at 21:00:00 (`(211122210000)`) your gas meter (`0-2:24.1.0(3)` --> type 3 = gas) reported 13368.864 M^3 (`(13368.864)` `(m3)`) ?

Yes it does.

**Comment by elucidator** [comment]
> I think I fixed it. I would really appreciate it if you could do some tests before I release it.
> 
> You can get version 0.6-SNAPSHOT from https://oss.sonatype.org/content/repositories/snapshots/

I verified with some message from my meter, and yes it look great.

**Comment by elucidator** [comment]
> Request: Please provide me a full message (or even better; 3 or 4 messages). I expect that after the '!' at the end there is a CRC code. If there is not then that is also something my parse must be able to deal with.

There is no additional data read no CRC code available.

Below a sample of multiple messages

```
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.712*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.713*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.715*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.717*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.718*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.720*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.722*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.59*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.723*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.60*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.725*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.58*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.727*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.58*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.728*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.57*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.730*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.58*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.731*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.58*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.733*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.57*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.735*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.66*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.737*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.67*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
/ISk5\2MT382-1003

0-0:96.1.1(5A424556303035313036383434393132)
1-0:1.8.1(16722.627*kWh)
1-0:1.8.2(19412.739*kWh)
1-0:2.8.1(00859.681*kWh)
1-0:2.8.2(01817.110*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.67*kW)
1-0:2.7.0(0000.00*kW)
0-0:17.0.0(0999.00*kW)
0-0:96.3.10(1)
0-0:96.13.1()
0-0:96.13.0()
0-2:24.1.0(3)
0-2:96.1.0(3238303131303038333036343239303133)
0-2:24.3.0(211123200000)(00)(60)(1)(0-2:24.2.1)(m3)
(13376.292)
0-2:24.4.0(1)
!
```

**Comment by nielsbasjes** [comment]
I released version 0.6 with these fixes in it.
